src/js.rs

```diff
@@ -0,0 +1,568 @@
+        TypedExpr::Tuple { elems, .. } => tuple(elems.iter().map(|e|
+            maybe_block_expr(e, &Env{return_last: &false})
+        )),
+        TypedExpr::TupleIndex { tuple, index, .. } => tuple_index(tuple, *index),
+
+        TypedExpr::Call { fun, args, .. } => call(fun, args),
+
+        // TODO is this always an anonymous fn
+        TypedExpr::Fn { args, body, .. } => fun(args, body),
+        
+        TypedExpr::Seq { first, then, .. } => seq(first, then, env),
+        TypedExpr::Var {
+            name, constructor, ..
+        } => var(name, constructor),
+        TypedExpr::Let {
+            value,
+            pattern,
+            then,
+            kind: BindingKind::Let,
+            ..
+        } => let_(value, pattern, then, env),
+        TypedExpr::BinOp {
+            name, left, right, ..
+        } => bin_op(name, left, right, &Env{return_last: &false}),
+        TypedExpr::Todo {
+            label, location, ..
+        } => {
+            println!("expression: {:?}", expression);
+
+            todo(label, *location)
+        },
+        _ => {
+            println!("expression: {:?}", expression);
+            unimplemented!("expr")
+        }
+    };
+    match expression {
+        // I would have thought let would be inside a sequence?
+        TypedExpr::Seq { .. } | TypedExpr::Let { .. } => "",
+        _ => match env.return_last {
+            true => "return ",
+            _ => ""
+        }
+    }.to_doc().append(rendered)
+}
+
+fn int<'a>(value: &str) -> Document<'a> {
+    Document::String(value.to_string())
+}
+
+fn float<'a>(value: &str) -> Document<'a> {
+    Document::String(value.to_string())
+}
+
+fn string(value: &str) -> Document<'_> {
+    value.to_doc().surround("\"", "\"")
+}
+
+fn expr_list_cons<'a>(head: &'a TypedExpr, tail: &'a TypedExpr) -> Document<'a> {
+    let env = Env{ return_last: &false};
+    let mut elements = vec![head];
+
+    let final_tail = collect_cons(tail, &mut elements);
+    let elements = elements.into_iter().map(|e| 
+        maybe_block_expr(e, &env)
+    );
+    let content = concat(Itertools::intersperse(elements, break_(",", ", ")));
+    let content = if let Some(final_tail) = final_tail {
+        content.append(Document::String(", ...".to_string()).append(maybe_block_expr(final_tail, &env)))
+    } else {
+        content
+    };
+    content.surround("[", "]")
+}
+
+fn collect_cons<'a>(tail: &'a TypedExpr, elements: &mut Vec<&'a TypedExpr>) -> Option<&'a TypedExpr> {
+    match tail {
+        TypedExpr::ListNil { .. } => None,
+        TypedExpr::ListCons { head, tail, .. } => {
+            elements.push(head);
+            collect_cons(tail, elements)
+        }
+        // TODO is it possible to write improper lists in Gleam
+        other => Some(other)
+    }
+}
+
+fn tuple<'a>(elems: impl Iterator<Item = Document<'a>>) -> Document<'a> {
+    concat(Itertools::intersperse(elems, break_(",", ", ")))
+        .nest_current()
```

Comment:
Can we indent normally rather than nest at the current level? I don't think JS uses that style of indentation.